File: packages/settlement/src/prizes.ts
/** Min–max normalise to [0,1]; if every value is equal the field is indistinguishable, so all get 0.5. */
export function minMaxNormalize(values: number[]): number[] {
  if (values.length === 0) return [];
  const min = Math.min(...values);
  const max = Math.max(...values);
  if (max === min) return values.map(() => 0.5);
  return values.map((v) => (v - min) / (max - min));
}

/** Largest peak-to-trough fall as a fraction of the peak (0 = never fell). */
export function maxDrawdown(series: number[]): number {
  let peak = -Infinity;
  let worst = 0;
  for (const v of series) {
    if (v > peak) peak = v;
    if (peak > 0) worst = Math.max(worst, (peak - v) / peak);
  }
  return worst;
}

/** Herfindahl–Hirschman index of holding weights (excludes cash). 1 = one holding, →0 = spread thin. */
export function herfindahl(holdingValues: number[]): number {
  const positive = holdingValues.filter((v) => v > 0);
  const total = positive.reduce((s, v) => s + v, 0);
  if (total <= 0) return 1;
  return positive.reduce((s, v) => s + (v / total) ** 2, 0);
}

export interface Prize1Input {
  fundId: string;
  /** Gross % NAV return over Phase 2, before fees. */
  navReturnPct: number;
  maxDrawdown: number;
  /** 0–1: share of the fund's own investors who ended Phase 2 in net profit. */
  investorProfitability: number;
  /** 0–1: capital retention through the reallocation windows (already floor-filtered, Sec 16). */
  retention: number;
}

export interface Prize1Weights {
  performance: number;
  riskManagement: number;
  investorProfitability: number;
  retention: number;
}

/**
 * Sec 16 Prize 1: Score = 100 × (0.35·Return + 0.25·RiskMgmt + 0.25·InvestorProfit
 * + 0.15·Retention), each component min–max normalised across the funds.
 * Risk management is the INVERSE of max drawdown. AUM is deliberately not an input.
 */
export function prize1Scores(funds: Prize1Input[], w: Prize1Weights) {
  const ret = minMaxNormalize(funds.map((f) => f.navReturnPct));
  const risk = minMaxNormalize(funds.map((f) => -f.maxDrawdown));
  const profit = minMaxNormalize(funds.map((f) => f.investorProfitability));
  const keep = minMaxNormalize(funds.map((f) => f.retention));

  return funds
    .map((f, i) => ({
      fundId: f.fundId,
      score:
        100 *
        (w.performance * ret[i] +
          w.riskManagement * risk[i] +
          w.investorProfitability * profit[i] +
          w.retention * keep[i]),
    }))
    .sort((a, b) => b.score - a.score)
    .map((f, i) => ({ ...f, rank: i + 1 }));
}

export interface Prize2Input {
  accountId: string;
  finalValue: number;
  /** Disqualified, or lost eligibility via the Sec 9 5%-minimum escalation. */
  eligible: boolean;
}

/** Sec 16 Prize 2: highest final portfolio value among eligible individual investors. */
export function prize2Ranking(entrants: Prize2Input[]) {
  return entrants
    .filter((e) => e.eligible)
    .sort((a, b) => b.finalValue - a.finalValue)
    .map((e, i) => ({ accountId: e.accountId, finalValue: e.finalValue, rank: i + 1 }));
}

export interface Prize4Input {
  accountId: string;
  returnPct: number;
  maxDrawdown: number;
  /** Market value of every non-cash holding, incl. fund units at NAV. */
  holdingValues: number[];
  eligible: boolean;
}

export interface Prize4Weights {
  riskAdjustedReturn: number;
  drawdownControl: number;
  diversification: number;
  diversificationCapHoldings: number;
}

/** Avoid dividing by ~0 when a portfolio never fell. */
const DRAWDOWN_FLOOR = 0.001;

/**
 * Sec 16 Prize 4 ("Capital Guardian"): risk-adjusted return (return ÷ max
 * drawdown) 40%, drawdown control 30%, diversification 30% — each min–max
 * normalised across eligible entrants. Diversification is 1/HHI (effective
 * number of holdings) capped at `diversificationCapHoldings` ("up to a
 * sensible point"), so past the cap more holdings stop helping.
 */
export function prize4Scores(entrants: Prize4Input[], w: Prize4Weights) {
  const pool = entrants.filter((e) => e.eligible);
  const riskAdj = minMaxNormalize(pool.map((e) => e.returnPct / Math.max(e.maxDrawdown, DRAWDOWN_FLOOR)));
  const control = minMaxNormalize(pool.map((e) => -e.maxDrawdown));
  const diversification = minMaxNormalize(
    pool.map((e) => {
      const hasHoldings = e.holdingValues.some((v) => v > 0);
      if (!hasHoldings) return 0;
      return Math.min(1 / herfindahl(e.holdingValues), w.diversificationCapHoldings);
    })
  );

  return pool
    .map((e, i) => ({
      accountId: e.accountId,
      score:
        100 *
        (w.riskAdjustedReturn * riskAdj[i] +
          w.drawdownControl * control[i] +
          w.diversification * diversification[i]),
    }))
    .sort((a, b) => b.score - a.score)
    .map((e, i) => ({ ...e, rank: i + 1 }));
}

File: apps/api/internal/scoring/prizes.go
package scoring

import (
	"math"
	"sort"

	"stockastic/api/internal/money"
	"stockastic/api/internal/rulebook"
)

// MinMaxNormalize maps values to [0,1]. If every value is equal the field cannot separate anyone,
// so all get 0.5.
func MinMaxNormalize(values []float64) []float64 {
	out := make([]float64, len(values))
	if len(values) == 0 {
		return out
	}
	lo, hi := values[0], values[0]
	for _, v := range values {
		lo, hi = math.Min(lo, v), math.Max(hi, v)
	}
	for i, v := range values {
		if hi == lo {
			out[i] = 0.5
		} else {
			out[i] = (v - lo) / (hi - lo)
		}
	}
	return out
}

// MaxDrawdown is the largest peak-to-trough fall as a fraction of the peak (0 = never fell).
func MaxDrawdown(series []float64) float64 {
	peak, worst := math.Inf(-1), 0.0
	for _, v := range series {
		if v > peak {
			peak = v
		}
		if peak > 0 {
			worst = math.Max(worst, (peak-v)/peak)
		}
	}
	return worst
}

// Herfindahl is the HHI of holding weights (cash excluded): 1 = a single holding, toward 0 = spread thin.
func Herfindahl(holdingValues []float64) float64 {
	var total float64
	for _, v := range holdingValues {
		if v > 0 {
			total += v
		}
	}
	if total <= 0 {
		return 1
	}
	var h float64
	for _, v := range holdingValues {
		if v > 0 {
			h += (v / total) * (v / total)
		}
	}
	return h
}

type Prize1Input struct {
	FundID string
	// NavReturnPct is gross % NAV return over Phase 2, before fees.
	NavReturnPct float64
	MaxDrawdown  float64
	// InvestorProfitability is 0-1: share of the fund's investors who ended in net profit.
	InvestorProfitability float64
	// Retention is 0-1, already filtered by the Sec 16 minimum-investment floor.
	Retention float64
}

type Scored struct {
	ID    string
	Score float64
	Rank  int
}

func rankScores(ids []string, scores []float64) []Scored {
	out := make([]Scored, len(ids))
	for i := range ids {
		out[i] = Scored{ID: ids[i], Score: scores[i]}
	}
	sort.SliceStable(out, func(i, j int) bool {
		if out[i].Score != out[j].Score {
			return out[i].Score > out[j].Score
		}
		return out[i].ID < out[j].ID
	})
	for i := range out {
		out[i].Rank = i + 1
	}
	return out
}

// Prize1Scores implements Sec 16 Prize 1: 100 x (w.perf*Return + w.risk*RiskMgmt + w.profit*InvestorProfit
// + w.retention*Retention), each component min-max normalised across the funds. Risk management is
// the INVERSE of max drawdown. AUM is deliberately not an input: size must not win.
func Prize1Scores(funds []Prize1Input, w rulebook.Prize1) []Scored {
	n := len(funds)
	ret, risk, profit, keep := make([]float64, n), make([]float64, n), make([]float64, n), make([]float64, n)
	ids := make([]string, n)
	for i, f := range funds {
		ids[i], ret[i], risk[i], profit[i], keep[i] = f.FundID, f.NavReturnPct, -f.MaxDrawdown, f.InvestorProfitability, f.Retention
	}
	nr, nk, np, nt := MinMaxNormalize(ret), MinMaxNormalize(risk), MinMaxNormalize(profit), MinMaxNormalize(keep)
	scores := make([]float64, n)
	for i := range funds {
		scores[i] = 100 * (w.Performance*nr[i] + w.RiskManagement*nk[i] + w.InvestorProfitability*np[i] + w.Retention*nt[i])
	}
	return rankScores(ids, scores)
}

type Prize2Input struct {
	AccountID  string
	FinalValue money.Paise
	// Eligible is false for disqualified teams and those that lost eligibility via Sec 9.
	Eligible bool
}

// Prize2Ranking is Sec 16 Prize 2: highest final portfolio value among eligible individual investors.
func Prize2Ranking(entrants []Prize2Input) []Scored {
	var in []Prize2Input
	for _, e := range entrants {
		if e.Eligible {
			in = append(in, e)
		}
	}
	ids, vals := make([]string, len(in)), make([]float64, len(in))
	for i, e := range in {
		ids[i], vals[i] = e.AccountID, float64(e.FinalValue)
	}
	return rankScores(ids, vals)
}

type Prize4Input struct {
	AccountID   string
	ReturnPct   float64
	MaxDrawdown float64
	// HoldingValues are the market values of every non-cash holding, including fund units at NAV.
	HoldingValues []float64
	// AvgEffectiveHoldings, if above zero, is the average spread of the investor's money over the whole event and is
	// used instead of HoldingValues, so diversifying only in the last minute cannot win the prize.
	AvgEffectiveHoldings float64
	Eligible             bool
}

// drawdownFloor avoids dividing by ~0 for a portfolio that never fell.
const drawdownFloor = 0.001

// Prize4Scores implements Sec 16 Prize 4 ("Capital Guardian"): risk-adjusted return (return / max
// drawdown), drawdown control, and diversification, each min-max normalised across eligible entrants.
// Diversification is the effective number of holdings 1/HHI, capped at DiversificationCapHoldings
// ("up to a sensible point").
func Prize4Scores(entrants []Prize4Input, w rulebook.Prize4) []Scored {
	var in []Prize4Input
	for _, e := range entrants {
		if e.Eligible {
			in = append(in, e)
		}
	}
	n := len(in)
	ids, riskAdj, control, div := make([]string, n), make([]float64, n), make([]float64, n), make([]float64, n)
	for i, e := range in {
		ids[i] = e.AccountID
		riskAdj[i] = e.ReturnPct / math.Max(e.MaxDrawdown, drawdownFloor)
		control[i] = -e.MaxDrawdown
		if e.AvgEffectiveHoldings > 0 {
			div[i] = math.Min(e.AvgEffectiveHoldings, w.DiversificationCapHoldings)
			continue
		}
		has := false
		for _, v := range e.HoldingValues {
			if v > 0 {
				has = true
			}
		}
		if has {
			div[i] = math.Min(1/Herfindahl(e.HoldingValues), w.DiversificationCapHoldings)
		}
	}
	nr, nc, nd := MinMaxNormalize(riskAdj), MinMaxNormalize(control), MinMaxNormalize(div)
	scores := make([]float64, n)
	for i := range in {
		scores[i] = 100 * (w.RiskAdjustedReturn*nr[i] + w.DrawdownControl*nc[i] + w.Diversification*nd[i])
	}
	return rankScores(ids, scores)
}
